#live processing with guppy R10

library(rhdf5)
source("~/nanocns/R-scripts/peroperative_sequencing_functions.R")
print("loaded")
#functioncheck_guppy()

#Fill in the fast5 output folder as generated by minKnow
fast5_folder="~/nanocns/data/PMC_20_1_fast5/" #to change, ends with "/fast5/"
#fast5_folder="~/nanocns/data/pmc_live_first_6_fast5s/"

#define an output directory
main_folder = "~/nanocns/data/PMC_live_20_1_redo_out/"
#specify a reference genome
refgenome="/home/sturgeon/nanocns/data/chm13_V2/chm13v2.0.fa"
#do you want the script to find the correct barcode (fill in: F), or do you want to override it (fill in the bc number)? 
bcoverride = 5
use_unclassified = TRUE

CNV_iterations = c(10,15)


system(paste0("mkdir " ,main_folder))
#functioncheck_guppy()

#work in progress leave in F for now
uselivesturgeon=F
#edit if you want to use smaller or bigger fast5 files (not recommended)
fast5_size = 4000

#don't change any of these
running = T
iteration = 1
progress = data.frame()
#When a megalodon instance is running, this file will be created and removed once megalodon is done. 
#If this file already exists, this script will wait until it is gone before starting a new megalodon, to prevent a system crash. 
#remove it manually, if no other version of this script is running or else it will wait forever.
if(file.exists("~/nanocns/wrapper_running.txt")){print("~/nanocns/wrapper_running.txt EXISTS!!")}

setwd(main_folder)
while(running == T){
  #start sturgeon if using live, this does not work right now
 
  #if(iteration==6){running=F}
  if(iteration==1){
    file.create("processed_files.txt")
    write("start",file="processed_files.txt",append = T )
  }
  #keep track of processed files
  processed_fast5s = read.table("processed_files.txt")
  made_fast5s = list.files(fast5_folder)
  #check if there are new files, if not, wait 30 seconds and check again
  new_fast5s = made_fast5s[made_fast5s%in%processed_fast5s$V1==F]
  
  if(length(new_fast5s)==0){
    print("no new files, waiting 30 seconds")
    Sys.sleep(30)
    next}
  #find out if they are ready for processing ie. if it contains 4000 reads
  fast5_cpy = c()
  for(fl in new_fast5s){
    tryCatch(len <- nrow(h5ls(file = paste0(fast5_folder,"/",fl), recursive=F)), error=function(x){len<-0})
    if(len==fast5_size){fast5_cpy=c(fast5_cpy,fl) }
    if(len!=0&len<fast5_size){print(paste0("incomplete file found: ",fl))}
  }
  #if it is not complete, wait 30 seconds and start over.
  #TODO why is this here twice??
  if(length(fast5_cpy)==0){
    print("no files to copy")
    Sys.sleep(30)
    next}
  #process files
  for(f5 in fast5_cpy){
    #make sure no wrapper is running already, this will overload the GPU RAM
    holdup = T
    while(holdup){
      if(file.exists("~/nanocns/wrapper_running.txt")){Sys.sleep(1)}else{holdup=F}
    }
    #make a file preventing someone else from starting a megalodon run
    system("touch ~/nanocns/wrapper_running.txt")
    
    print(paste0("starting wrapper for file: ", f5))
    #wrapper starts guppy and extracts reads from the correct barcode
    wrappert_guppy_R10(main_folder, paste0(fast5_folder,"/", f5), iteration = iteration, bcoverride = bcoverride, include_unclassified = use_unclassified)
    
    system("rm ~/nanocns/wrapper_running.txt")
    
    iteration=iteration+1
    
    write(f5, file="processed_files.txt",append=T)
    #add iteration results to the progress dataframe, and make a plot
    progress = try(add_and_plot(progress, paste0(main_folder, "/merged_probes_methyl_calls_model.csv")), silent=T)
    Sys.sleep(5)
    if(iteration%in%CNV_iterations){
      pdf(file=paste0(main_folder,"/CNV_plot_iteration",iteration,".pdf"),useDingbats = F)
      try(plot_cnv_from_live_dir_DNAcopy(main_folder), silent=T)
      dev.off()
      try(system(paste0("open ", main_folder,"/CNV_plot_iteration",iteration,".pdf")), silent=T)
    Sys.sleep(5)}
  }
}

#end sturgeon
system("pkill -f sturgeon")


